server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/IssueCreationDateCalculator.java

```diff
@@ -0,0 +1,123 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.task.projectanalysis.issue;
+
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.Optional;
+import org.sonar.api.utils.DateUtils;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.core.issue.DefaultIssue;
+import org.sonar.core.issue.IssueChangeContext;
+import org.sonar.server.computation.task.projectanalysis.analysis.Analysis;
+import org.sonar.server.computation.task.projectanalysis.analysis.AnalysisMetadataHolder;
+import org.sonar.server.computation.task.projectanalysis.component.Component;
+import org.sonar.server.computation.task.projectanalysis.qualityprofile.ActiveRule;
+import org.sonar.server.computation.task.projectanalysis.qualityprofile.ActiveRulesHolder;
+import org.sonar.server.computation.task.projectanalysis.scm.Changeset;
+import org.sonar.server.computation.task.projectanalysis.scm.ScmInfo;
+import org.sonar.server.computation.task.projectanalysis.scm.ScmInfoRepository;
+import org.sonar.server.issue.IssueFieldsSetter;
+
+import static org.sonar.core.issue.IssueChangeContext.createScan;
+
+/**
+ * Calculates the creation date of an issue. Takes into account, that the issue
+ * might be raised by adding a rule to a quality profile.
+ */
+public class IssueCreationDateCalculator extends IssueVisitor {
+
+  private static final Logger LOGGER = Loggers.get(IssueCreationDateCalculator.class);
+
+  private final ScmInfoRepository scmInfoRepository;
+  private final IssueFieldsSetter issueUpdater;
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private final IssueChangeContext changeContext;
+  private final ActiveRulesHolder activeRulesHolder;
+
+  public IssueCreationDateCalculator(AnalysisMetadataHolder analysisMetadataHolder, ScmInfoRepository scmInfoRepository,
+    IssueFieldsSetter issueUpdater, ActiveRulesHolder activeRulesHolder) {
+    this.scmInfoRepository = scmInfoRepository;
+    this.issueUpdater = issueUpdater;
+    this.analysisMetadataHolder = analysisMetadataHolder;
+    this.changeContext = createScan(new Date(analysisMetadataHolder.getAnalysisDate()));
+    this.activeRulesHolder = activeRulesHolder;
+  }
+
+  @Override
+  public void onIssue(Component component, DefaultIssue issue) {
+    if (issue.isNew() && ruleIsNew(issue)) {
+
+      // get scm information of file, if available
```

Comment:
hint of this being abusive: 
when there is no changeset, date of issue is not altered... this is hidden in the last line `ifPresent` while for understanding by reader, this should be written as a fail fast condition

functional programming is about making code closer to the algorithm... this is quite the opposite